### ISSUE TITLE
Attempting to read outside of stack

### ISSUE THREAD

**Issue by aristotelhs** [DataDog/ddprof#276]
Hello, 
we are trying to use ddprof for our continuous profiling. Currently we are deploying ddprof as a sidecar container that is used to profile the main process. The main process is written in Julia and it is using the Julia jit for the execution of the end code. 
Checking the result in the web interface, we can see that some frames are missing. Running ddprof in debug mode we notice a lot of error messages like:
```
<DEBUG>Jun 11 12:07:19.028064 ddprof[143]: Attempting to read outside of stack 0x7ffd39f95258 from 47, (0x7ffd39f8d0d0, 0x7ffd39f950d0)[0x7f418fb376b8, 0x7f418fb3f6b8]
<DEBUG>Jun 11 12:07:19.028067 ddprof[143]: Attempting to read outside of stack 0x7ffd39f95260 from 47, (0x7ffd39f8d0d0, 0x7ffd39f950d0)[0x7f418fb376b8, 0x7f418fb3f6b8]
<DEBUG>Jun 11 12:07:19.028070 ddprof[143]: Attempting to read outside of stack 0x7ffd39f95268 from 47, (0x7ffd39f8d0d0, 0x7ffd39f950d0)[0x7f418fb376b8, 0x7f418fb3f6b8]
<DEBUG>Jun 11 12:07:19.028073 ddprof[143]: Attempting to read outside of stack 0x7ffd39f95278 from 47, (0x7ffd39f8d0d0, 0x7ffd39f950d0)[0x7f418fb376b8, 0x7f418fb3f6b8]
<DEBUG>Jun 11 12:07:19.083331 ddprof[143]: Attempting to read outside of stack 0x7ffd39f92878 from 47, (0x7ffd39f8a730, 0x7ffd39f92730)[0x7f418faff790, 0x7f418fb07790]
<DEBUG>Jun 11 12:07:19.083354 ddprof[143]: Attempting to read outside of stack 0x7ffd39f928a0 from 47, (0x7ffd39f8a730, 0x7ffd39f92730)[0x7f418faff790, 0x7f418fb07790]
<DEBUG>Jun 11 12:07:19.083359 ddprof[143]: Attempting to read outside of stack 0x7ffd39f92880 from 47, (0x7ffd39f8a730, 0x7ffd39f92730)[0x7f418faff790, 0x7f418fb07790]
<DEBUG>Jun 11 12:07:19.083362 ddprof[143]: Attempting to read outside of stack 0x7ffd39f92888 from 47, (0x7ffd39f8a730, 0x7ffd39f92730)[0x7f418faff790, 0x7f418fb07790]
```
Also in the last part we get errors like this:

```
<DEBUG>Jun 11 12:28:33.665010 ddprof[143]: [DSO] Couldn't find DSO for [47](0x7f08633164ae). backpopulate
<DEBUG>Jun 11 12:28:33.665038 ddprof[143]: [UW] (PID47) DSO not found at 0x7f08633164ae (depth#15)
<DEBUG>Jun 11 12:28:33.665043 ddprof[143]: Error frame (depth#16)
<DEBUG>Jun 11 12:28:33.665047 ddprof[143]: [DSO] Couldn't find DSO for [47](0x7f0863316631). backpopulate
<DEBUG>Jun 11 12:28:33.665049 ddprof[143]: [UW] (PID47) DSO not found at 0x7f0863316631 (depth#16)
<DEBUG>Jun 11 12:28:33.665052 ddprof[143]: Error frame (depth#17)
<DEBUG>Jun 11 12:28:33.665676 ddprof[143]: [DSO] Couldn't find DSO for [47](0x7f08633164ae). backpopulate
<DEBUG>Jun 11 12:28:33.665684 ddprof[143]: [UW] (PID47) DSO not found at 0x7f08633164ae (depth#24)
<DEBUG>Jun 11 12:28:33.665686 ddprof[143]: Error frame (depth#25)
```

 I also include the pid maps for the process. 
[47-map.txt](https://github.com/DataDog/ddprof/files/11714640/47-map.txt)

Is there a way to say ddprof to continue unwinding the stack even in cases that this is a non stack segment? There are legit reasons that this would happen during to code generation (and it points to executable maps in the file).

**Comment by aristotelhs** [comment]
If I read the first comment correctly this means that even I get this error message about the address being out of stack range this means that you continue the unwinding using FP. (We have compiled Julia to use FP so this should work as expected). 

The out of bounds read is in different place that the second part of the log messages. So how did you correlate the out of bounds with the size of the sample? 

For us it is important to have the full stacks and it seems that we do end up with deep stacks because there is a process of compiling our DSL to Julia and then to JIT. Not being able to go deeper in the stack we might loose some of the code paths that lead to specific execution. (@NHDaly might be better to give an insight on this).  If in the 32kb samples the stack unwinding is exhausted , are  the relevant stack traces send to DD, or they are dropped?  (Trying to see how important the error messages are).

**Comment by NHDaly** [comment]
+1 it's hard for us to answer without understanding those details.

If the samples are dropped, that's bad and we won't like it.
If the samples are sent and somehow can indicate that they're truncated, that would be more acceptable. Would they be truncated at the leaves or the roots? Or rather, which part would be preserved?

Obviously we'd most prefer to have the full stacks, but we could probably make do with partial stacks if it's our only option. ... I'm not sure if it would be enough of a detriment to prevent us from using the service or not. We'd have to see.

**Comment by r1viollet** [comment]
Question around what parts are being kept: The leaves are kept, the roots are dropped (so frames closer to your `main` are dropped, whereas the leaf functions like `memcpy` are kept). But the sample as a whole is kept.

We capture a sample with the state in which the code is being executed and then we make our way up to the calling functions. 
When we are unable to complete the unwinding, you will see an `[incomplete]` frame explaining that we were not able to fully unwind.

**Comment by NHDaly** [comment]
aha, thanks, we had wondered what that meant. Makes sense! So `[incomplete]` indicates some number of frames that were dropped?

----

> Higher samples would have more impact on your runtime (to be measured). We could make this into a setting.

Yes, a setting sounds great! I think our preference would likely be:
- dramatically lower the sample frequency (we currently have it set to 19Hz i think(?), rather than the 100Hz default)
- but increase the sample size so that we can keep the entire frame

With those two knobs, we should be able to keep the perf impact neutral, while retaining full profile information.

Right now, our application performance tuning questions are much more in the "why did this take 10 seconds" domain, than in the "why did this take 10 microseconds", so we strongly prefer full profiles over fine-grained samples.

**Comment by NHDaly** [comment]
But, with all of this said: the above discussion still doesn't cover the scenario we were seeing, which seemed to be presenting as _dropped samples,_ right?

Is your expectation that nothing we've shown indicates this is happening, @r1viollet?

@aristotelhs: can you run an experiment where you trigger high CPU load on an engine, and then check to see whether or not the profile reflects that high CPU usage as expected?

**Comment by r1viollet** [comment]
> Is your expectation that nothing we've shown indicates this is happening, @r1viollet?
> 
Yes, this is correct.

The only scenario in which we would be loosing samples is if we are unable to keep up with the number of samples  or if we are not scheduled to process the samples. I can check this if you connect the profiler to a statsd socket (not billed in any way). You would need to mount the agent's statsd socket in the container and provide it to the profiler

`--internal_stats /var/run/datadog-agent/statsd.sock`

**Comment by aristotelhs** [comment]
So `[Incomplete]` is when we reach the maximum depth of the stack unwinding. In a previous message I thought that Incomplete was created because there was a version mismatch on the libraries making unwinding fail, or can it be both? In our case I can see that almost all traces are Incomplete.

We run ddprof with debug (-l debug)  so when I run the load I get this output:
```
<NOTICE>Jun 18 09:41:06.537646 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:42:05.464464 ddprof[143]: datadog.profiling.native.event.lost: 8
<NOTICE>Jun 18 09:43:04.463343 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:44:03.568167 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:45:02.454830 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:46:01.468762 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:47:00.567233 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:47:59.553123 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:48:58.544933 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:49:57.527206 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:50:56.510943 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:51:55.459613 ddprof[143]: datadog.profiling.native.event.lost: 6
<NOTICE>Jun 18 09:52:54.543147 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:53:53.533412 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:54:52.553264 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:55:51.493461 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:56:50.503270 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:57:49.525502 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:58:48.547727 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 09:59:47.484624 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 10:00:46.546879 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 10:01:45.461844 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 10:02:44.513070 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 10:03:43.508614 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 10:04:42.537522 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 10:05:41.467133 ddprof[143]: datadog.profiling.native.event.lost: 0
<NOTICE>Jun 18 10:06:40.551488 ddprof[143]: datadog.profiling.native.event.lost: 0
```
However in the test run that I do I am able to see some profiles coming from the test setup. (Not sure if something is missing). I will leave it running so that we can have a better look with @NHDaly .

**Comment by r1viollet** [comment]
That looks like a healthy amount of lost events. The number of samples processed should also be in the logs.

I would recommend the `notice` setting which will give you general metrics on how the profiler is behaving while not suffering form some of the performance impacts of the `debug` logs.
`[Incomplete]` is when we don't detect a frame that matches what we would expect from a successful unwinding . Here are symbols we expect at the end of the unwinding:
-      "__clone"
-     "__clone3"
-     "_exit"
-     "main"
-     "runtime.goexit.abi0"
-     "runtime.systemstack.abi0"
-     "_start"
-     "start_thread"

This can be due to the unwinding reaching the end of the buffer we sampled, or some missing information (such as missing unwinding or symbol information). Incomplete frames are usually a minor inconvenience (as the leaf nodes are the most important) and should not prevent you interpreting performance results.

**Comment by NHDaly** [comment]
Thanks for the support so far. This is really helpful!

We spent some time looking into this today, and we do have some strange findings still.

Here is one example of the inconsistent CPU reporting between the metrics and the profiles:
- [datadog metric `kubernetes.cpu.user.total`](https://app.datadoghq.com/metric/explorer?start=1687532160000&end=1687532460000&paused=true#N4Ig7glgJg5gpgFxALlAGwIYE8D2BXJVEADxQEYAaELcqyKBAC1pEbghkcLIF8qo4AMwgA7CAgg4RKUAiwAHOChASAtnADOcAE4RNIKtrgBHPJoQaUAbVBGN8qVoD6gnNtUZCKiOq279VKY6epbINiAiGOrKQdpYZAYgUJ4YThr42gDGSsgg6gi6mZaBZnHKGABuMMgA1ngARjoiiJoAdJnyeK14fq0IOAgYaMDaGBCtcCIwonBOkerIjSKZjAC02vWCq54WGqt7mRjJAMxkqwAM57wgPAC6VK7ueJih4Y+qzxgxpfE39yAaORoHKgeRDRAIHJJHAwJyZF4aCCZRJoGZOOSKZTpVFQFFo+hMZQiNweNA3fgQeyYLDohRQ1HNP48PgA+SohAAYSkwhgKBEzzQPCAA)
    <img width="447" alt="Screenshot 2023-06-23 at 10 00 39 AM" src="https://github.com/DataDog/ddprof/assets/1582097/73d2467a-e344-49d3-adaa-64a553efdd26">
- [datadog cpu profiles for the same pod, and the same time window](https://app.datadoghq.com/profiling/search?query=rai.account_name%3Arelationalai-team-rd-julia-eng%20rai.engine_name%3Abench-rbf-attss-s-cada31-001&agg_m=%40prof_native_cpu_cores_by_top_methods.value&agg_q=%40prof_native_cpu_cores_by_top_methods.attributes.method&agg_t=avg&viz=stream&start=1687532160000&end=1687532460000&paused=true):
    <img width="956" alt="Screenshot 2023-06-23 at 10 03 10 AM" src="https://github.com/DataDog/ddprof/assets/1582097/2d4a1b06-c3d1-4898-9bef-656ff6517061">


The pod has 4 cores. The metrics dashboard reports somewhere between 1 and 3 cores in use during those 5 minutes.
But the ddprof profile reports only up to 1.0 core in use.
Shouldn't the ddprof profile get all the way up to 3+ cores, too?

Is something happening that is normalizing the profiles by number of cores? This is not expected, from my understanding.

Thanks in advance for the support!

**Comment by NHDaly** [comment]
All the uploads during that window report the same ([logs](https://app.datadoghq.com/logs?query=-status%3Adebug%20datadog.profiling.native.event.lost%20%40rai.engine_name%3Abench-rbf-attss-s-cada31-001&cols=host%2Cservice&event=AgAAAYjowf9sQjf4KQAAAAAAAAAYAAAAAEFZam93Z3BHQUFDRGxSQ3dUZDZQaFFBaQAAACQAAAAAMDE4OGU4ZTktNDU2OS00Njg2LTk5MTMtZmE5Y2M1MmFhMmNh&index=%2A&messageDisplay=inline&stream_sort=desc&viz=stream&from_ts=1687532160000&to_ts=1687532460000&live=false)):
```
<NOTICE>Jun 23 14:56:47.839272 ddprof[40905]: datadog.profiling.native.event.lost: 0
```

**Comment by NHDaly** [comment]
I do see this debug message over and over - not sure if it's an issue?:
```
<DEBUG>Jun 23 16:14:44.640740 ddprof[40905]: Stopped at 7ea7f1489f43 - dso PID[7] 7ea7f1400000-7ea7f14fffff 0 ()(T-Anonymous)(x)(ID#-1) - error no error ()
<DEBUG>Jun 23 16:14:44.640759 ddprof[40905]: Break out of unwinding (possible infinite loop)
```

**Comment by NHDaly** [comment]
Finally, I wanted to point out that often the samples are pretty uselessly short right now, e.g.:
<img width="790" alt="Screenshot 2023-06-23 at 10 28 26 AM" src="https://github.com/DataDog/ddprof/assets/1582097/1b9b9030-8caf-407a-bebb-0c34343e4b6c">


Do you think that we could increase the buffer length, as you suggested? Also, can you remind me what `Anonymous` means? It seems to correlate with the JITted frames.

**Comment by aristotelhs** [comment]
Regarding the discussion about the Incomplete, I can notice that Incomplete exists in all of our traces. So after the top level entry point that is the executable the next top level trace is [Incomplete]. This means that our traces end up on something different of the ones that you expect from ddprof.

**Comment by NHDaly** [comment]
+1 all of our profiles have incomplete at the top.

**Comment by r1viollet** [comment]
> Thanks for the support so far. This is really helpful!
> 
> We spent some time looking into this today, and we do have some strange findings still.
> 
> ....
> 
> The pod has 4 cores. The metrics dashboard reports somewhere between 1 and 3 cores in use during those 5 minutes. But the ddprof profile reports only up to 1.0 core in use. Shouldn't the ddprof profile get all the way up to 3+ cores, too?
> 
> Is something happening that is normalizing the profiles by number of cores? This is not expected, from my understanding.
> 
> Thanks in advance for the support!

We should be reporting the full CPU activity. Here is one of our services running at ~4 CPUs per minute.
![image](https://github.com/DataDog/ddprof/assets/74836499/039df01c-8f99-4dd0-891b-b0179506ad54)

I can not reproduce this on my side. Would you have any hints as to how I could create an example exposing such an issue ? Are you confident that the activity on that POD is only explained by this process ?

**Comment by r1viollet** [comment]
To recap what I understand. What we are sure about:
- From some of the logs you added above, the symbolization flow can be improved. 
I made a small PR to improve the flow.

- The size of samples are too small, we should make this configurable.
This is the most likely cause for most of the unwinding issues we are seeing. You might need to change some ulimit settings to make the new samples size work. We use pinned memory to create the buffers. This is a kernel setting.

The remaining item that I find strange is the CPU accounting not matching what you would expect it to be.

**Comment by aristotelhs** [comment]
Thanks for the reply. 
The way we run ddprof is that we inject the ddprof container which when starts monitors the julia process by running ddprof -p $PID_OF_JULIA_PROCESS.
In the same pod it is the julia process, ddprof itself and a small container that is used to expose a socket from the host. The cpu limits exist only in the Julia process and not to the other pods.
When you saw the native-prof-analyzer process when you go to the right side of the window and you choose Thread-ID how many does it show? In our julia process when we go and ask to show by Thread-ID we only get one Thread-ID back. 

Which are the new ulimits that we must add? Should we wait for the patch to be merged before we try?

**Comment by NHDaly** [comment]
> Are you confident that the activity on that POD is only explained by this process ?

I am confident, yes. We are running julia with multiple threads, and we know for sure that it can utilize multiple cores.

> Would you have any hints as to how I could create an example exposing such an issue ?

Maybe we can get together, and pair on this sometime? It would be great to get your support if that's something you're able to do. I think we can get you the permissions so you can trigger workloads on our side to support your debugging!

**Comment by NHDaly** [comment]
Thanks both! +1 to the discussion above

**Comment by aristotelhs** [comment]
Quick question: If we deploy ddprof as a daemonset will it suffer from the same issue regarding threads or is it using a different code path?  (Having done the test yet, but it is good to have this in mind)

**Comment by r1viollet** [comment]
No, the daemonset using global mode won't suffer from this. This only relates to PID mode.

**Comment by r1viollet** [comment]
A new options is available in the pre-release version. It will increase the size of samples we take. You can find the binaries [here](https://github.com/DataDog/ddprof/releases/tag/v0.13.0-rc). I will soon release a new version of `ddprof`.
The command line should like like this:
```
./ddprof --stack_sample_size=65528 -l notice --preset cpu_only julia main.jl
```
Can you confirm that this improves the stack traces ?

**Comment by r1viollet** [comment]
The stack size confguration was released to v0.13.0

**Comment by aristotelhs** [comment]
For the discussion on the incomplete profiles, Julia seems to have the start_task symbols that is used to label top level tasks. So this might improve some traces. 
I will upgrade to v0.13.0 to see the changes. We tried the container-id option which seems to be able to isolate the container, but we cannot group both on container-id and pid. 
The container-id is useful but we need an easy way to do the mapping to pod -> container-ids since most people are used to think in pods rather than container-ids.
Thanks for the new release.
FYI we use ddprof as a daemonset that runs in parallel to our apps

**Comment by aristotelhs** [comment]
The logs from the new containers are the following:
```<NOTICE>Jul 20 18:36:55.903177 ddprof[1710354]: Increasing size order of the ring buffer to 8 (from 6)
<ERROR>Jul 20 18:36:55.904327 ddprof[1710354]: Could not mmap memory for watcher #0: Operation not permitted. Please increase kernel limits on pinned memory (ulimit -l). OR associate the IPC_LOCK capability to this process.
<ERROR>Jul 20 18:36:55.904339 ddprof[1710354]: PERFMMAP: error in mmap operations at /go/src/github.com/DataDog/apm-reliability/ddprof-build/ddprof/src/pevent_lib.cc:188
<ERROR>Jul 20 18:36:55.904343 ddprof[1710354]: Forward error at /go/src/github.com/DataDog/apm-reliability/ddprof-build/ddprof/src/pevent_lib.cc:227 - PERFMMAP: error in mmap operations
<NOTICE>Jul 20 18:36:55.904369 ddprof[1710354]: Retrying attachment without user override```

The logs are with the IPC_LOCK capability added to the containers. I have added the `--stack_sample_size=65528` to the running command. Is this parameter ignored then?

**Comment by r1viollet** [comment]
**Filtering:**

Thanks for this feedback. I have forwarded it. This requires rethinking some of the filtering mechanisms.
I will get back to you on these improvements.

**Mmap log failure:**

This is mostly a confusing log. We will retry mmaps with different users (because limitations are shared across containers for a given user). The first user failed, and we are now trying with the second user (still with a larger buffer). I'll look at how to make this less confusing.

**Comment by r1viollet** [comment]
As mentioned, the logs is misleading, the parameter is still taken into account. Could you clarify if the flamegraphs are deep enough ?

**Comment by aristotelhs** [comment]
Hello, using this parameter we do have a better outcome for the stack depth. As can be seen in the attached picture.
![image](https://github.com/DataDog/ddprof/assets/48688196/309e5070-28b3-41dc-b265-f4296e6b5e92)
The [Incomplete] part has shrug significantly.  However in that part it looks strange because I can also see frames that do not look that deep as shown in the next pic:
![image](https://github.com/DataDog/ddprof/assets/48688196/66a4447a-e5eb-4b9e-abba-7f7d1d66f003)
The frames routed  ijl_gc_* functions do not seem to be that deep to explain of being [Incomplete]. I am not sure if this is a case of being routed in some "fake" or special stack frames that results in the unwinding issues of showing as incomplete. (cc @vchuravy). 
The anonymous frames are a very small percentage in this run (~0,25%) and they got improved by adding some debuginfo packages in our builds. Is there an easy way to see which debug symbols are missing, or if there is some mismatch on the C libraries to remove them? 
![image](https://github.com/DataDog/ddprof/assets/48688196/55083888-732f-48b4-9491-7659a22056f6)
One thing that I noticed in the logs are the following lines:
```
<NOTICE>Aug 04 10:33:57.087393 ddprof[2853012]: datadog.profiling.native.symbols.jit.reads: 2
<NOTICE>Aug 04 10:33:57.087394 ddprof[2853012]: datadog.profiling.native.symbols.jit.failed_lookups: 377
<NOTICE>Aug 04 10:33:57.087395 ddprof[2853012]: datadog.profiling.native.symbols.jit.symbol_count: 0
```
Does that mean that ddprof is unable to read the jitdump files for the processes?  In the traces I do see symbols in the format julia_sortNOT._18923 which look like jitted functions.

We run ddprof as a pod in the cluster nodes and we monitor all the processes. The command line arguments are the following: `ddprof -l notice --stack_sample_size=65528 -e sCPU,period=19 -g`

**Comment by aristotelhs** [comment]
Running some extra loads we get the following profile.

<img width="1145" alt="image" src="https://github.com/DataDog/ddprof/assets/48688196/cd4dec53-a7ab-4fa1-8f50-6ba3b65f9f78">

It seems the jitted code is not shown and the stack frames are missing. It looks like all the Anonymous frames that exist in the profile are the jitted code. I am not sure why they appear much less deep that they actual are ( I would expect to have the correct depth and on each level the function name would be anonymous, but it seems that an aggregation is happening).

**Comment by NHDaly** [comment]
^ I don't think this is quite right. I don't think we know whether the Anonymous frames are the jitted code or not, but we know for sure that stacks with the jitted frames are completely missing.

We ran the above workload in both julia's native profiler and in DDProf, and you can see that whole samples are missing from the profile, which seem to correspond to samples with jitted frames:
![Screenshot 2023-08-04 at 11 30 51 AM](https://github.com/DataDog/ddprof/assets/1582097/7ba15e2f-6c50-46df-9efe-975eb496a448)
![Screenshot 2023-08-04 at 11 30 45 AM](https://github.com/DataDog/ddprof/assets/1582097/fad80f47-8761-4b68-b60e-b76d37772e57)

In particular, the main column on the left, the `run_job` column, is completely absent. 

---

and regarding the question of rooting: notice that the `sort` columns _are_ rooted in `start_task` in the julia native profile, and they're reasonably short, so I don't see why those should end at incomplete.

**Comment by r1viollet** [comment]
Thanks for the logs and screen captures. 

The most important issues seems to be that the JITDump lookups are failing. I am working on a reproducer and I will see how to fix this.
You highlighted other things that might need some attention: 
- Unwinding through signal handlers (looks like Julia has a segv handler)
- Stack depth still not being deep enough.

Though we should come back to that once we have proper JIT symbols in the flamegraphs.

**Comment by r1viollet** [comment]
I think this can relate to how jitdump files are specified in the maps
```
.debug/jit/llvm-IR-jit-20230611-ec3bc0/jit-47.dump
```
Is this relative to the julia installation ?

I was expecting absolute paths
```
/home/r1viollet/.debug/jit/llvm-IR-jit-20230824-dc3310/jit-1499363.dump
```

**Comment by r1viollet** [comment]
As a workaround, can you try specifying the path to the jitdump folder explicitly ?
```
export JITDUMPDIR=<path_where_the_debug_folder_will_be_created>
```

**Comment by aristotelhs** [comment]
> 

If I read the code correctly this changes the directory where julia (through llvm) would create the debug dir. In our case we are in the case of home dir (https://github.com/llvm/llvm-project/blob/main/llvm/lib/ExecutionEngine/PerfJITEvents/PerfJITEventListener.cpp#L306) so we just add the . in front.
If we change JITDUMPDIR in the julia process how would ddprof catch that and find the directory?

**Comment by aristotelhs** [comment]
If there is a docker image with the version of ddprof with the above patch, I can try to install that and test it in our infra.

**Comment by r1viollet** [comment]
From my tests with `JITDUMPDIR`, an absolute path was generated (you can check this in the /proc/<julia_process_pid>/maps, the line with the .dump file is either absolute or relative).

I will have a patch version you can try by end of day.

**Comment by aristotelhs** [comment]
I am not sure this is the case because I just spinned up a new engine and then when I checked the maps file I do get the full path:
```
bash-5.1$ grep dump /proc/1/maps
7f8e53749000-7f8e5374a000 r-xp 00000000 00:ff 4644907                    /home/user/.debug/jit/llvm-IR-jit-20230805-bbab0f/jit-1.dump
```
I also created a new ddprof binary with the patches in `r1viollet/jitdump_relative_path` and if ddprof could not find the jitdump file then as I read it it should emit a log (WARN) that it is unable to open the jitdump file, which is not the case in our infra.

**Comment by r1viollet** [comment]
Indeed, I added a warning (from the latest patch). That occurs if we were able to find the .dump mapping.
The following [release](https://github.com/DataDog/ddprof/releases/tag/v0.14.1-rc) contains the patch.

Though you were able to run your own versions :clap: .
We can coordinate on Datadog public slack if that is OK to debug this together, I am not sure why we would not find this.

**Comment by r1viollet** [comment]
The latest 0.14.1 release includes a patch that allows the `--global` option to match the jitdump information.
@aristotelhs was able to confirm the improvement.

We also noticed an issue with the unwinding in the Julia segv handler (used by the runtime within the garbage collector). We will discuss this in a separate issue.

Thanks for the help investigating this issue.
